File: app/src/main/java/com/pr0gramm/app/parcel/LambdaCreator.java
package com.pr0gramm.app.parcel;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.common.base.Function;

/**
 */
public class LambdaCreator<T extends Parcelable> implements Parcelable.Creator<T> {
    private final Function<Parcel, T> createFromParcel;
    private final Function<Integer, T[]> newArray;

    public LambdaCreator(Function<Parcel, T> createFromParcel, Function<Integer, T[]> newArray) {
        this.createFromParcel = createFromParcel;
        this.newArray = newArray;
    }

    @Override
    public T createFromParcel(Parcel source) {
        return createFromParcel.apply(source);
    }

    @Override
    public T[] newArray(int size) {
        return newArray.apply(size);
    }
}

File: app/src/main/java/com/pr0gramm/app/parcel/LambdaCreator.kt
package com.pr0gramm.app.parcel

import android.os.Parcel
import android.os.Parcelable

/**
 */
class LambdaCreator<T : Parcelable>(
        private val fnCreateFromParcel: (Parcel) -> T,
        private val fnNewArray: (Int) -> Array<T>) : Parcelable.Creator<T> {

    override fun createFromParcel(source: Parcel): T {
        return fnCreateFromParcel(source)
    }

    override fun newArray(size: Int): Array<T> {
        return fnNewArray(size)
    }
}
